src/common/glob.h

```diff
@@ -0,0 +1,158 @@
+        // search early without trying to match them to longer
+        // substrings. This is because a longer match for the
+        // earlier part of the pattern would require the rest of the
+        // pattern to match starting later in the string, and we
+        // have just determined that there is no match for the rest
+        // of the pattern starting from anywhere in the current
+        // string.
+        *skip_longer_matches = true;
+        return false;
+      case '?':
+        if (string.empty()) return false;
+        string.remove_prefix(1);
+        break;
+      case '[': {
+        pattern.remove_prefix(1);
+        const bool invert = pattern[0] == '^';
+        if (invert) pattern.remove_prefix(1);
+
+        bool match = false;
+        while (true) {
+          if (pattern.empty()) {
+            // unterminated [ group: reject invalid pattern
+            return false;
+          } else if (pattern[0] == ']') {
+            break;
+          } else if (pattern.length() >= 2 && pattern[0] == '\\') {
+            pattern.remove_prefix(1);
+            if (pattern[0] == string[0]) match = true;
+          } else if (pattern.length() >= 3 && pattern[1] == '-') {
+            unsigned char start = canonicalize(pattern[0]);
+            unsigned char end = canonicalize(pattern[2]);
+            if (start > end) std::swap(start, end);
+            const int c = canonicalize(string[0]);
+            pattern.remove_prefix(2);
+
+            if (c >= start && c <= end) match = true;
+          } else if (canonicalize(pattern[0]) == canonicalize(string[0])) {
+            match = true;
+          }
+          pattern.remove_prefix(1);
+        }
+        if (invert) match = !match;
+        if (!match) return false;
+        string.remove_prefix(1);
+        break;
+      }
+      case '\\':
+        if (pattern.length() >= 2) {
+          pattern.remove_prefix(1);
+        }
+        [[fallthrough]];
+      default:
+        // Just a normal character
+        if (!ignore_case) {
+          if (pattern[0] != string[0]) return false;
+        } else {
+          if (std::tolower((int)pattern[0]) != std::tolower((int)string[0])) return false;
+        }
+        string.remove_prefix(1);
+        break;
+    }
+    pattern.remove_prefix(1);
+  }
+
+  // Now that either the pattern is empty or the string is empty, this is a match iff
+  // the pattern consists only of '*', and the string is empty.
+  return string.empty() && std::all_of(pattern.begin(), pattern.end(), [](char c) { return c == '*'; });
+}
+
+// Given a glob [pattern] and a string [string], return true iff the string matches the glob.
+// If [ignore_case] is true, the match is case-insensitive.
+constexpr inline bool GlobMatches(std::string_view glob, std::string_view str, bool ignore_case = false) {
+  bool skip_longer_matches = false;
+  return GlobMatchImpl(glob, str, ignore_case, &skip_longer_matches);
+}
```

Comment:
Are we able to reuse the glob impl in string_util.h? If not it's fine to impl it here.

Reply:
Thanks, I changed `StringMatch` and added some performance optimizations and guards against exponential complexity attacks `KEYS *?*?*?*?*?*?*?*...` which used to completely max out the CPU usage.